x/auth/spec/README.md

```diff
@@ -18,27 +32,846 @@ are performed, and exposes the account keeper, which allows other modules to rea
+// a notion of account numbers for replay protection for previously pruned accounts,
+// and a pubkey for authentication purposes.
+//
+// Many complex conditions can be used in the concrete struct which implements AccountI.
+type AccountI interface {
+	proto.Message
+
+	GetAddress() sdk.AccAddress
+	SetAddress(sdk.AccAddress) error // errors if already set.
+
+	GetPubKey() crypto.PubKey // can return nil.
+	SetPubKey(crypto.PubKey) error
+
+	GetAccountNumber() uint64
+	SetAccountNumber(uint64) error
+
+	GetSequence() uint64
+	SetSequence(uint64) error
+
+	// Ensure that account implements stringer
+	String() string
+}
+```
+
+##### Base Account
+
+A base account is the simplest and most common account type, which just stores all requisite
+fields directly in a struct.
+
+```protobuf
+// BaseAccount defines a base account type. It contains all the necessary fields
+// for basic account functionality. Any custom account type should extend this
+// type for additional functionality (e.g. vesting).
+message BaseAccount {
+  string address = 1;
+  google.protobuf.Any pub_key = 2;
+  uint64 account_number = 3;
+  uint64 sequence       = 4;
+}
+```
+
+#### Vesting Account
+
+See [Vesting](#vesting).
+
+## 04. Messages
+
+The auth module does not expose any transaction message.
+
+### Handlers
+
+The auth module presently has no transaction handlers of its own, but does expose
+the special `AnteHandler`, used for performing basic validity checks on a transaction,
+such that it could be thrown out of the mempool. Note that the ante handler is called on
+`CheckTx`, but _also_ on `DeliverTx`, as Tendermint proposers presently have the ability
+to include in their proposed block transactions which fail `CheckTx`.
+
+#### Ante Handler
+
+```go
+anteHandler(ak AccountKeeper, fck FeeCollectionKeeper, tx sdk.Tx)
+  if !tx.(StdTx)
+    fail with "not a StdTx"
+
+  if isCheckTx and tx.Fee < config.SubjectiveMinimumFee
+    fail with "insufficient fee for mempool inclusion"
+
+  if tx.ValidateBasic() != nil
+    fail with "tx failed ValidateBasic"
+
+  if tx.Fee > 0
+    account = GetAccount(tx.GetSigners()[0])
+    coins := acount.GetCoins()
+    if coins < tx.Fee
+      fail with "insufficient fee to pay for transaction"
+    account.SetCoins(coins - tx.Fee)
+    fck.AddCollectedFees(tx.Fee)
+
+  for index, signature in tx.GetSignatures()
+    account = GetAccount(tx.GetSigners()[index])
+    bytesToSign := StdSignBytes(chainID, acc.GetAccountNumber(),
+      acc.GetSequence(), tx.Fee, tx.Msgs, tx.Memo)
+    if !signature.Verify(bytesToSign)
+      fail with "invalid signature"
+
+  return
+```
+
+## 05. Keepers
+
+The auth module only exposes one keeper, the account keeper, which can be used to read and write accounts.
+
+### Account Keeper
+
+Presently only one fully-permissioned account keeper is exposed, which has the ability to both read and write
+all fields of all accounts, and to iterate over all stored accounts.
+
+```go
+type AccountKeeper interface {
```

Comment:
This interface doesn't seem to exist anymore in x/auth.

2 notes:
- there is a `type AccountKeeper interface` in most modules, I think the one in x/auth just got deleted?
- godoc mentions a `sdk.AccountKeeper`

https://github.com/cosmos/cosmos-sdk/blob/bf8809ef9840b4f5369887a38d8345e2380a567f/x/auth/keeper/account.go#L8-L9

but I can't find that either.

I am going to add back a `type AccountKeeper interface` in x/auth. But if anyone has any other idea, please let me know.